### PR TITLE
optimization of access level table construction

### PR DESCRIPTION
**Pull request by Bryanskiy**
Refactoring which was mentioned in #87487

### CODE REVIEW

**Comment by fee1-dead** [comment]
Thanks for the pull request! I am going to request this to be reviewed by someone else because I am not familiar enough with that part of the code.

r? rust-lang/compiler

File: compiler/rustc_resolve/src/access_levels.rs [hunk 1]
```diff
@@ -75,30 +71,20 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
-                .cloned()
-                .collect::<Vec<_>>();
-
-            let module = self.r.get_module(module_id.to_def_id()).unwrap();
-            for export in pub_exports.into_iter() {
-                if let Some(export_def_id) = export.res.opt_def_id().and_then(|id| id.as_local()) {
-                    self.set_access_level_def_id(export_def_id, Some(AccessLevel::Exported));
-                }
+        let module = self.r.get_module(module_id.to_def_id()).unwrap();
+        let resolutions = self.r.resolutions(module).clone();
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:75]: Cloning the resolution table is an expensive operation, you have to borrow it in some way.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 2]
```diff
@@ -183,14 +169,6 @@ impl<'r, 'ast> Visitor<'ast> for AccessLevelsVisitor<'ast, 'r> {
...
                 self.prev_level = orig_level;
             }
-
-            ast::ItemKind::ForeignMod(ForeignMod { ref items, .. }) => {
-                for nested in items {
-                    if nested.vis.kind.is_pub() {
-                        self.set_access_level(nested.id, access_level);
-                    }
-                }
-            }
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:193]: `fn set_exports_access_level` still doesn't visit items removed here and below, so they are no longer visited.

`fn set_exports_access_level` should either visit all items and not only imports (probably preferable long term), or this code needs to be restored.

File: compiler/rustc_resolve/src/lib.rs [hunk 3]
```diff
@@ -2006,6 +2006,23 @@ impl<'a> Resolver<'a> {
         }
         self.main_def = Some(MainDefinition { res, is_import, span });
     }
+
+    pub fn export_check(&self, binding: &NameBinding<'a>) -> Option<def::Res<!>> {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/lib.rs:2010]:
```suggestion
    pub fn is_reexport(&self, binding: &NameBinding<'a>) -> Option<def::Res<!>> {
```
this also needs a comment, something like
```
// Items that go to reexport table encoded to metadata and visible through it to other crates.
```

File: compiler/rustc_resolve/src/access_levels.rs [hunk 4]
```diff
@@ -72,33 +70,60 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
+                            | DefKind::Trait
+                            | DefKind::TraitAlias
+                            | DefKind::AssocFn
+                            | DefKind::AssocConst
+                            | DefKind::AssocTy
+                            | DefKind::Macro(..) => {
+                                if binding.vis.is_public() {
+                                    self.set_access_level_def_id(def_id, self.prev_level);
+                                } else {
+                                    self.set_access_level_def_id(def_id, None);
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:111]: `set_access_level_def_id` cannot decrease the level, so this line is always noop.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 5]
```diff
@@ -72,33 +70,60 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
+        let module = self.r.get_module(module_id.to_def_id()).unwrap();
+        let resolutions = self.r.resolutions(module);
 
-        if let Some(exports) = self.r.reexport_map.get(&module_id) {
-            let pub_exports = exports
-                .iter()
-                .filter(|ex| ex.vis == Visibility::Public)
-                .cloned()
-                .collect::<Vec<_>>();
+        for (.., name_resolution) in resolutions.borrow_mut().iter() {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:76]:
```suggestion
        for (.., name_resolution) in resolutions.borrow().iter() {
```

File: compiler/rustc_resolve/src/access_levels.rs [hunk 6]
```diff
@@ -127,70 +152,44 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
-            | ast::ItemKind::Trait(..)
-            | ast::ItemKind::TraitAlias(..)
-            | ast::ItemKind::MacroDef(..) => {
-                if item.vis.kind.is_pub() {
-                    self.prev_level
+                if item.attrs.iter().any(|attr| attr.has_name(sym::macro_export)) {
+                    self.set_access_level(item.id, Some(AccessLevel::Public));
                 } else {
-                    None
+                    self.set_access_level(item.id, None);
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:179]: This line is also always noop, see the previous comment.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 7]
```diff
@@ -127,70 +152,44 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
-            ast::ItemKind::ForeignMod(ForeignMod { ref items, .. }) => {
-                for nested in items {
-                    if nested.vis.kind.is_pub() {
-                        self.set_access_level(nested.id, access_level);
-                    }
-                }
-            }
             ast::ItemKind::Enum(EnumDef { ref variants }, _) => {
                 for variant in variants {
                     let variant_level = self.set_access_level(variant.id, access_level);
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:195]: Constructors in variants and structures don't need to be visited if they are processed in `fn set_exports_access_level` (which at this point can be renamed to `fn set_bindings_access_level`).

File: compiler/rustc_resolve/src/access_levels.rs [hunk 8]
```diff
@@ -214,13 +213,8 @@ impl<'r, 'ast> Visitor<'ast> for AccessLevelsVisitor<'ast, 'r> {
                     }
                 }
             }
-            ast::ItemKind::Trait(ref trait_kind) => {
-                for nested in trait_kind.items.iter() {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:218]: Traits and enums are also "modules" from name resolution point of view.
So you can use `self.set_exports_access_level(self.r.local_def_id(item.id));` in `ItemKind::Enum`s and `ItemKind::Trait`s, like in `ItemKind::Mod`.

Without that items inside traits are not visited at all if this code is removed.

File: compiler/rustc_resolve/src/access_levels.rs (hunk 7 above)
Comment (petrochenkov) [reply compiler/rustc_resolve/src/access_levels.rs:195]: In the end we should only visit fields, `ForeignMod`s, and `MacroDef`s in `fn visit_item`, everything else will be done using `NameBinding` visiting.

File: compiler/rustc_feature/src/builtin_attrs.rs [hunk 9]
```diff
@@ -762,6 +762,7 @@ pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
     // Internal attributes, Testing:
     // ==========================================================================
 
+    rustc_attr!(TEST, rustc_access_levels, Normal, template!(NameValueStr: "value"), WarnFollowing),
```
[the remark this answers is not in the document]
Comment (Bryanskiy) [reply compiler/rustc_feature/src/builtin_attrs.rs:765]: `span_err` is not available due to `#![cfg_attr(not(bootstrap), deny(rustc::untranslatable_diagnostic))]`. I use `delay_span_bug` instead which call `panic!`. In that case testing in `src/test/ui` becomes impossible without specifying access level due to backtraces.

[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_feature/src/builtin_attrs.rs:765]: You can (preferably) create an error structure for this error, like they are created in `compiler\rustc_privacy\src\errors.rs` for other errors in `compiler\rustc_privacy\src\lib.rs`.
If you don't want to do that, you can just put `#[allow(rustc::untranslatable_diagnostic)]` on the function using `span_err`.

[the remark this answers is not in the document]
Comment (Bryanskiy) [reply compiler/rustc_feature/src/builtin_attrs.rs:765]: As I understand, `compiler\rustc_privacy\src\errors.rs` is used for user errors, but not for internal diagnostic. I suggest `span error` use without error structure.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 10]
```diff
@@ -127,99 +124,59 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
 
 impl<'r, 'ast> Visitor<'ast> for AccessLevelsVisitor<'ast, 'r> {
     fn visit_item(&mut self, item: &'ast ast::Item) {
-        let inherited_item_level = match item.kind {
+        let mut access_level = match item.vis.kind.is_pub() {
+            true => self.prev_level,
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:128]: The `prev_level` field is unnecessary and can be removed.
The same information can be obtained by using `self.r.access_levels.map.get(...)` on the parent `LocalDefId`.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 11]
```diff
@@ -127,99 +124,59 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
+                if item.attrs.iter().any(|attr| attr.has_name(sym::macro_export)) {
+                    self.set_access_level(item.id, Some(AccessLevel::Public));
+                }
+            }
 
-        // Set access level of nested items.
-        // If it's a mod, also make the visitor walk all of its items
-        match item.kind {
             ast::ItemKind::Mod(..) => {
+                access_level = self.set_access_level(item.id, access_level);
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:153]: This line is unnecessary, this level should already be set by parent module's `set_bindings_access_level`.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 12]
```diff
@@ -72,33 +69,33 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
+        let resolutions = self.r.resolutions(module);
 
-        if let Some(exports) = self.r.reexport_map.get(&module_id) {
-            let pub_exports = exports
-                .iter()
-                .filter(|ex| ex.vis == Visibility::Public)
-                .cloned()
-                .collect::<Vec<_>>();
+        for (.., name_resolution) in resolutions.borrow().iter() {
+            if let Some(binding) = name_resolution.borrow().binding() {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:76]: The logic in this block can be simplified a lot, to something like
```rust
                if binding.vis.is_public() && !binding.is_ambiguity() {
                    set_import_binding_access_level(self, binding, module_level);
                    if let Some(def_id) = binding.res().opt_def_id().and_then(|id| id.as_local()) {
                        self.set_access_level_def_id(
                            def_id,
                            if binding.is_import() {
                                Some(AccessLevel::Exported)
                            } else {
                                module_level
                            },
                        );
                    }
                }
```
or maybe even more if `set_import_binding_access_level` is inlined.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 13]
```diff
@@ -72,33 +69,33 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
+                    if let Res::Def(_, _) = binding.res() {
+                        if binding.vis.is_public() {
+                            self.set_access_level_def_id(def_id, self.prev_level);
+                        }
                     }
                 }
             }
         }
     }
-
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:101]: Nit: no whitespace between items.

File: src/test/ui/privacy/access_levels.rs [hunk 14]
```diff
@@ -0,0 +1,50 @@
+#![feature(rustc_attrs)]
+#![allow(unused)]
```
Comment (petrochenkov) [inline src/test/ui/privacy/access_levels.rs:2]: Warnings from the `unused` group are allowed by default in UI tests, so this attribute is unnecessary.

File: compiler/rustc_privacy/src/lib.rs [hunk 15]
```diff
@@ -904,6 +905,60 @@ impl<'tcx> DefIdVisitor<'tcx> for ReachEverythingInTheInterfaceVisitor<'_, 'tcx>
...
+    tcx: TyCtxt<'tcx>,
+    access_levels: &'tcx AccessLevels,
+}
+
+impl<'tcx> TestReachabilityVisitor<'tcx> {
+    fn access_level_diagnostic(&mut self, def_id: LocalDefId) {
+        if self.tcx.has_attr(def_id.to_def_id(), sym::rustc_access_level) {
+            let access_level = format!("{:?}", self.access_levels.map.get(&def_id));
+            let span = self.tcx.def_span(def_id.to_def_id());
+            self.tcx.sess.emit_err(ReportAccessLevel { span: span, descr: access_level });
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:921]:
```suggestion
            self.tcx.sess.emit_err(ReportAccessLevel { span, descr: access_level });
```
Field init shorthand can be used here.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 16]
```diff
@@ -44,13 +39,12 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
 
     fn reset(&mut self) {
         self.changed = false;
-        self.prev_level = Some(AccessLevel::Public);
     }
 
     /// Update the access level of the exports of the given module accordingly. The module access
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:44]:
```suggestion
    /// Update the access level of the bindings in the given module accordingly. The module access
```

File: compiler/rustc_resolve/src/access_levels.rs [hunk 17]
```diff
@@ -123,100 +112,75 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
             old_level
         }
     }
+
+    fn get_prev_access_level(&mut self, def_id: LocalDefId) -> Option<AccessLevel> {
+        match self.r.opt_local_parent(def_id) {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:117]:
```suggestion
        self.r.local_parent(def_id) {
```
`ast::Item` always has a parent (this function can be inlined and removed).

File: compiler/rustc_resolve/src/access_levels.rs [hunk 18]
```diff
@@ -123,100 +112,75 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
+            None => Some(AccessLevel::Public),
+        }
+    }
 }
 
 impl<'r, 'ast> Visitor<'ast> for AccessLevelsVisitor<'ast, 'r> {
     fn visit_item(&mut self, item: &'ast ast::Item) {
-        let inherited_item_level = match item.kind {
+        let def_id = self.r.local_def_id(item.id);
+        let prev_level = self.get_prev_access_level(def_id);
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:127]:
```suggestion
        let parent_level = self.get_prev_access_level(def_id);
```

File: compiler/rustc_resolve/src/access_levels.rs [hunk 19]
```diff
@@ -123,100 +112,75 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
             }
 
-            ast::ItemKind::ForeignMod(ForeignMod { ref items, .. }) => {
-                for nested in items {
-                    if nested.vis.kind.is_pub() {
-                        self.set_access_level(nested.id, access_level);
-                    }
+            ast::ItemKind::Mod(..) => {
+                let current_level = self.r.access_levels.map.get(&def_id);
+                if current_level.is_some() {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:157]: It's simpler to do this check on `module_level` inside `set_bindings_access_level`.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 20]
```diff
@@ -123,100 +112,75 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
+                    self.set_bindings_access_level(def_id);
                 }
+                visit::walk_item(self, item);
             }
+
             ast::ItemKind::Enum(EnumDef { ref variants }, _) => {
                 for variant in variants {
                     let variant_level = self.set_access_level(variant.id, access_level);
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:165]: Now you are (incorrectly) setting the level for variants, but not for variant constructors.
Could you use `set_bindings_access_level` here as I suggested in https://github.com/rust-lang/rust/pull/100147/files#r944705524?

File: compiler/rustc_resolve/src/access_levels.rs [hunk 21]
```diff
@@ -123,100 +112,75 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
             }
-            ast::ItemKind::Struct(ref def, _) | ast::ItemKind::Union(ref def, _) => {
-                if let Some(ctor_id) = def.ctor_id() {
-                    self.set_access_level(ctor_id, access_level);
-                }
 
+            ast::ItemKind::Struct(ref def, _) | ast::ItemKind::Union(ref def, _) => {
                 for field in def.fields() {
                     if field.vis.kind.is_pub() {
                         self.set_access_level(field.id, access_level);
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:175]: Field level should be inherited from the struct/union itself, not from its parent.

**Comment by petrochenkov** [comment]
r=me after squashing commits (2 commits, probably, because `#[rustc_access_level]` and `compiler/rustc_resolve/src/access_levels.rs` refactoring are two distinct things).
@rustbot author

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Pull request merged**